### ISSUE TITLE
faster mean, variance for computing dispersion

### ISSUE THREAD

**Issue by fidelram** [scverse/scanpy#163]
I did a time comparison on the scanpy function to compute mean and variance and the sklearn function that does the same.

Here some results with a very large matrix (84.7 million values,):

```python
adata.X
```
> <8847360x58051 sparse matrix of type '<class 'numpy.float32'>'
> 	with 84702299 stored elements in Compressed Sparse Row format>
> 

**Using current implementation**:
```python
import scanpy.preprocessing.simple as simple
%timeit simple._get_mean_var(adata.X)
```
2.3 s ± 114 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

**Using sklearn sparse functions**:
```python
import sklearn.utils.sparsefuncs as sparsefuncs
def unbiased_estimator(X):
    mean, var =sparsefuncs.mean_variance_axis(X, 0)
    # enforce R convention (unbiased estimator) for variance
    var *= (X.shape[0]/(X.shape[0]-1))
    return mean, var
%timeit unbiased_estimator(adata.X)
```
> 141 ms ± 6.95 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
>

The results returned by both methods are different only in the precision:

```python
# the variance of both methods
simple._get_mean_var(adata.X)[1], unbiased_estimator(adata.X)[1]
```
> (array([0.0073651 , 0.00061313, 0.005346  , ..., 0.05514137, 0.01749513,
>         0.05664281], dtype=float32),
>  array([0.0073651 , 0.00061313, 0.00534599, ..., 0.05514137, 0.01749513,
>         0.05664282], dtype=float32))
> 

```python
np.allclose(simple._get_mean_var(adata.X)[1], unbiased_estimator(adata.X)[1], rtol=1e-4,)
```
> True

This comparison is for the variance, but for the mean, the same results are obtained

I suggest to replace the `_get_mean_var` function for the sklearn function. **Note** the sklearn function only works with a sparse matrix.

**Comment by falexwolf** [comment]
Very nice! Thank you for the benchmark! 🙂 It's a bit astonishing as I investigated this some a bit more than a year ago but maybe there was some improvement...

I am in principle happy to change this soon - I don't think that the change within floating point precision will change results [it does, for instance, in PCA... of course not a qualitative result, but nonetheless a tSNE could be rotated etc.]!

How did you come to investigating this? Was the computation a bottleneck for you? The matrix you provide is almost - for current standards - "unrealistically large".

**Comment by fidelram** [comment]
I was looking at the scanpy function to compute the mean and variance an noticed that it had some comments inside, pointing to performance issues. Thus, I looked for an alternative method, found the sklearn sparse function and then tested it in an artificially large matrix. Otherwise, I did not have any trouble with the current implementation.

The floating point precision is higher in the sklearn method, thus I suppose this is not an issue.

**Comment by ivirshup** [comment]
Kinda related, I was about to open an issue on the memory usage of this function. The current implementation can double the memory usage of a program, I believe due to intermediate arrays in the current code.

I'd come up with a slower but fewer allocation method for dense arrays (which could probably be sped up with a little `numba`):

```python
def lessalloc_dense(X):
    mean = X.mean(axis=0)
    mean_sq = np.apply_along_axis(lambda x: np.square(x).mean(), 0, X)
    var = (mean_sq - mean**2) * ((X.shape[0]/(X.shape[0]-1)))
    return mean, var
```

And looked at memory usage using [`memory_profiler`](https://github.com/pythonprofilers/memory_profiler/releases), including @fidelram 's method:

![mean_var_memory](https://user-images.githubusercontent.com/8238804/40597918-eacd8764-6287-11e8-98ff-017e697b350d.png)

[Full script for benchmark here.](https://gist.github.com/ivirshup/a6facfa1ace5b356ea2d18ff3ffe0cb9)

**Comment by ivirshup** [comment]
Fixed in #857
